yt_dlp/extractor/substack.py

```diff
@@ -0,0 +1,98 @@
+import re
+
+from .common import InfoExtractor
+
+
+class SubstackIE(InfoExtractor):
+    _VALID_URL = r'https?://(?P<username>[\w-]+)\.substack\.com/p/(?P<id>[\w-]+)'
+    _TESTS = [{
+        'url': 'https://haleynahman.substack.com/p/i-made-a-vlog?s=r',
+        'md5': 'f27e4fc6252001d48d479f45e65cdfd5',
+        'info_dict': {
+            'id': '47660949',
+            'ext': 'mp4',
+            'title': 'I MADE A VLOG',
+            'description': 'md5:10c01ff93439a62e70ce963b2aa0b7f6',
+            'thumbnail': 'md5:bec758a34d8ee9142d43bcebdf33af18',
+            'uploader': 'haleynahman',
+        }
+    }, {
+        'url': 'https://haleynahman.substack.com/p/-dear-danny-i-found-my-boyfriends?s=r',
+        'md5': '0a63eacec877a1171a62cfa69710fcea',
+        'info_dict': {
+            'id': '51045592',
+            'ext': 'mpga',
+            'title': "🎧 Dear Danny: I found my boyfriend's secret Twitter account",
+            'description': 'md5:a57f2439319e56e0af92dd0c95d75797',
+            'thumbnail': 'md5:daa40b6b79249417c14ff8103db29639',
+            'uploader': 'haleynahman',
+        }
+    }, {
+        'url': 'https://andrewzimmern.substack.com/p/mussels-with-black-bean-sauce-recipe',
+        'md5': 'fd3c07077b02444ff0130715b5f632bb',
+        'info_dict': {
+            'id': '47368578',
+            'ext': 'mp4',
+            'title': 'Mussels with Black Bean Sauce: Recipe of the Week #7',
+            'description': 'md5:b96234a2906c7d854d5229818d889515',
+            'thumbnail': 'md5:e30bfaa9da40e82aa62354263a9dd232',
+            'uploader': 'andrewzimmern',
+        }
+    }]
+
+    @staticmethod
+    def is_custom_domain(webpage):
+        return re.search(r'<script[^>]+src=["\']https://substackcdn.com/.*\.js', webpage) is not None
+
+    @staticmethod
+    def url_from_custom_domain(url):
+        custom_domain = r'https?://(?:www\.)?(?P<username>[\w-]+)\.(?:[a-z]+)/(?P<params>.+)'
+
+        username = re.search(custom_domain, url).group('username')
+        params = re.search(custom_domain, url).group('params')
+
+        return f'https://{username}.substack.com/{params}'
```

Comment:
like this

```suggestion
    @classmethod
    def _extract_url(cls, webpage, url):
        if not re.search(r'<script[^>]+src=["\']https://substackcdn.com/[^"\']+\.js', webpage):
            return
        parsed = urlparse(url)
        return parsed._replace(netloc=f'{parsed.netloc.split(".")[-2]}.substack.com').geturl()
```

Reply:
Do we actually have a guarantee that a Substack custom domain corresponds to an identically-spelt subdomain of `substack.com`?

Reply:
I just noticed we can get the subdomain from the page, can you check this commit? https://github.com/yt-dlp/yt-dlp/pull/4011/commits/2a563e100949122ea6a8c40a73c6138ddbab3cd2

Reply:
Okay, I like that more. Personally, I would have just used the custom domain during extraction, but this looks acceptable too.

I guess I would have liked something less (overtly) regexp-based as well, but it’s not like I have a better solution at the ready.